pkg/util/log/gophercloud.go

```diff
@@ -0,0 +1,73 @@
+package log
+
+import (
+	"net/http"
+	"strings"
+	"time"
+
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+)
+
+func NewLoggingProviderClient(endpoint string, logger kitlog.Logger) (*gophercloud.ProviderClient, error) {
+	providerClient, err := openstack.NewClient(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	providerClient.HTTPClient.Transport = &loggingRoundTripper{
+		&http.Transport{},
+		logger,
+	}
+
+	return providerClient, err
+}
+
+type loggingRoundTripper struct {
+	rt     http.RoundTripper
+	Logger kitlog.Logger
+}
+
+func (lrt *loggingRoundTripper) RoundTrip(request *http.Request) (response *http.Response, err error) {
+	logger := kitlog.With(lrt.Logger,
+		"method", request.Method,
+		"host", request.URL.Hostname(),
+		"path", request.URL.EscapedPath(),
+	)
+
+	query := request.URL.Query()
+	for i, k := range query {
+		logger = kitlog.With(logger,
+			i, strings.Join(k, ","))
+	}
+
+	defer func(begin time.Time) {
+		if response != nil {
+			logger = kitlog.With(logger,
+				"status", response.StatusCode,
+				"size", response.ContentLength,
```

Comment:
why is this always -1? i suppose it needs to be wrapped into a buffer or something.

Reply:
From the go docs:
> ContentLength records the length of the associated content. The
>  value -1 indicates that the length is unknown. Unless Request.Method
>  is "HEAD", values >= 0 indicate that the given number of bytes may
>  be read from Body.

I think the server (openstack service) doesn't set the content-length header as this would require buffering the whole response before sending anything.